share/dkg/dkg.go

```diff
@@ -0,0 +1,611 @@
+		d.reconstructed[rs.DealerIndex] = true
+		delete(d.pendingReconstruct, rs.DealerIndex)
+	}
+	return nil
+}
+
+// Finished returns true if the DKG has operated the protocol correctly and has
+// all necessary information to generate the DistKeyShare() by itself. It
+// returns false otherwise.
+func (d *DistKeyGenerator) Finished() bool {
+	var ret = true
+	var nb = 0
+	d.qualIter(func(idx uint32, v *vss.Verifier) bool {
+		nb++
+		// ALL QUAL members should have their commitments by now either given or
+		// reconstructed.
+		if _, ok := d.commitments[idx]; !ok {
+			ret = false
+			return false
+		}
+		return true
+	})
+	return nb >= d.t && ret
+}
+
+// ProduceSharedSecret generates the sharedsecret relative to this receiver
+// It throws an error if something is wrong such as not enough deals received.
+// The shared secret can be computed when all deals have been sent and
+// basically consists of a
+// 1. Public point which is the sum of all aggregated individual public commits
+// of each individual secrets.
+// 2. Share of the global Private Polynomial, basically SUM of fj(i) for a receiver i
+func (d *DistKeyGenerator) DistKeyShare() (*DistKeyShare, error) {
+	if !d.Certified() {
+		return nil, errors.New("dkg: distributed key not certified")
+	}
+
+	sh := d.suite.Scalar().Zero()
+	var pub *share.PubPoly
+	var err error
+
+	d.qualIter(func(i uint32, v *vss.Verifier) bool {
+		// share of dist. secret = sum of all share received.
+		s := v.Deal().SecShare.V
+		sh = sh.Add(sh, s)
+		// Dist. public key = sum of all revealed commitments
+		poly, ok := d.commitments[i]
+		if !ok {
+			err = fmt.Errorf("dkg: protocol not finished: %d commitments missing", i)
+			return false
+		}
+		if pub == nil {
+			// first polynomial we see (instead of generating n empty commits)
+			pub = poly
+			return true
+		}
+		pub, err = pub.Add(poly)
+		if err != nil {
+			return false
+		}
+		return true
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	_, commits := pub.Info()
+
+	return &DistKeyShare{
+		Commits: commits,
+		Share: &share.PriShare{
+			I: int(d.index),
+			V: sh,
+		},
+	}, nil
+}
+
+func (sc *SecretCommits) Hash(s abstract.Suite) []byte {
+	h := s.Hash()
+	h.Write([]byte("secretcommits"))
+	binary.Write(h, binary.LittleEndian, sc.Index)
+	for _, p := range sc.Commitments {
+		p.MarshalTo(h)
+	}
+	return h.Sum(nil)
+}
+
+func (cc *ComplaintCommits) Hash(s abstract.Suite) []byte {
+	h := s.Hash()
+	h.Write([]byte("commitcomplaint"))
+	binary.Write(h, binary.LittleEndian, cc.Index)
+	binary.Write(h, binary.LittleEndian, cc.DealerIndex)
+	h.Write(cc.Deal.Hash(s))
+	return h.Sum(nil)
+}
+
+func (rc *ReconstructCommits) Hash(s abstract.Suite) []byte {
+	h := s.Hash()
+	h.Write([]byte("reconstructcommits"))
+	binary.Write(h, binary.LittleEndian, rc.Index)
+	binary.Write(h, binary.LittleEndian, rc.DealerIndex)
+	h.Write(rc.Share.Hash(s))
+	return h.Sum(nil)
+}
+
+// XXX: maybe put that as internal package for vss & dkg since they both use the
```

Comment:
There's no decision on this as I'm not sure what's the best thing to do here.
On one hand, it's only one function and quite short and completely "independant" (does not need to call any other functions).
On the other hand it's duplicated code...